### ISSUE TITLE
fugazzi - Intrinsic arbitrage due to price feed discrepancies could lead to loss of value for the protocol

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-02-rio-network-core-protocol-judging#358]
fugazzi

high

# Intrinsic arbitrage due to price feed discrepancies could lead to loss of value for the protocol

## Summary

Users of the protocol can profit from the intrinsic arbitrage opportunities caused by potential Chainlink price discrepancies when converting to the unit of account.

## Vulnerability Detail

LRTs in the Rio protocol support multiple assets. When users deposit, the selected asset and amount is first converted to a unit of account that is normalized for all assets. This normalization is done using Chainlink price oracles.

```solidity
/// @notice Converts an asset amount to its equivalent value in the unit of account. The unit of
/// account is the price feed's quote asset.
/// @param asset The address of the asset to convert.
/// @param amount The amount of the asset to convert.
function convertToUnitOfAccountFromAsset(address asset, uint256 amount) public view returns (uint256) {
    if (asset == ETH_ADDRESS) {
        return amount;
    }
    address priceFeed = assetInfo[asset].priceFeed;
    uint256 price = getPrice(priceFeed);

    return _normalizeDecimals(price * amount / priceScale, assetInfo[asset].decimals, priceFeedDecimals);
}
```

In a similar way, when users request a withdrawal, they choose an asset and their share of the LRT token is converted from the unit of account to the asset.

```solidity
/// @notice Converts the unit of account value to its equivalent in the asset. The unit of
/// account is the price feed's quote asset.
/// @param asset The address of the asset to convert to.
/// @param value The asset's value in the unit of account.
function convertFromUnitOfAccountToAsset(address asset, uint256 value) public view returns (uint256) {
    if (asset == ETH_ADDRESS) {
        return value;
    }
    address priceFeed = assetInfo[asset].priceFeed;
    uint256 price = getPrice(priceFeed);

    return _normalizeDecimals(value * priceScale / price, priceFeedDecimals, assetInfo[asset].decimals);
}
```

Chainlink's price feed updates are triggered by different conditions. They have a heartbeat and a deviation threshold, and each configuration can potentially vary by feed. For example, a feed with a deviation threshold of 2% means that the feed price may be off by up to 2% with respect to the _real_ tracked price (the price has moved but it hasn't been updated onchain).

Any price discrepancy between assets creates intrinsic arbitrage opportunities that could be exploited in detriment of the protocol. Since deposits can be done using any of the underlying assets, a user could deposit one asset and request a withdrawal using another asset which has a pending on-chain price update. For example, if token B price is lagged with a pending +~2% price update, a user could deposit using token A (which would be considered more expensive) and then request a withdrawal using token B (considered cheaper), and get away with more value than originally deposited.

Even if all of the underlying assets in a LRT are expected to be similar assets and follow a similar price reference (for example, using multiple ETH LSTs that follow the price of ETH), their respective Chainlink price feeds may not be aligned at all times.

## Impact

Bad actors can take advantage of these discrepancies and extract value from the protocol by arbitraging between different underlying assets.

There is an excellent analysis with on-chain data as evidence in the [following twitter thread](https://twitter.com/m_Rassska/status/1743709612799341056).

As reference, [the following report](https://github.com/code-423n4/2023-11-kelp-findings/issues/584) also depicts the same issue described here.

## Code Snippet

https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/main/rio-sherlock-audit/contracts/restaking/RioLRTCoordinator.sol#L77-L88

https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/main/rio-sherlock-audit/contracts/restaking/RioLRTCoordinator.sol#L99-L116

## Tool used

Manual Review

## Recommendation

Given the presence of multiple underlying assets, it is dangerous to solely rely on Chainlink oracles given the way these operate. A potential mitigation could factor different oracles, such as a TWAP to ease out these discrepancies.

Duplicate of #386
